llvm/lib/CodeGen/MachineSSAUpdater2.cpp

```diff
@@ -0,0 +1,180 @@
+  while (!LiveInBlockWorklist.empty()) {
+    MachineBasicBlock *BB = LiveInBlockWorklist.pop_back_val();
+
+    // The block really is live in here, insert it into the set.  If already in
+    // the set, then it has already been processed.
+    if (!LiveInBlocks.insert(BB).second)
+      continue;
+
+    // Since the value is live into BB, it is either defined in a predecessor or
+    // live into it to.  Add the preds to the worklist unless they are a
+    // defining block.
+    for (MachineBasicBlock *P : BB->predecessors()) {
+      // The value is not live into a predecessor if it defines the value.
+      if (DefBlocks.count(P))
+        continue;
+
+      // Otherwise it is, add to the worklist.
+      LiveInBlockWorklist.push_back(P);
+    }
+  }
+}
+
+MachineInstrBuilder
+MachineSSAUpdater2::createInst(unsigned Opc, MachineBasicBlock *BB,
+                               MachineBasicBlock::iterator I) {
+  return BuildMI(*BB, I, DebugLoc(), TII.get(Opc),
+                 MRI.createVirtualRegister(RegAttrs));
+}
+
+// IsLiveOut indicates whether we are computing live-out values (true) or
+// live-in values (false).
+Register MachineSSAUpdater2::computeValue(MachineBasicBlock *BB,
+                                          bool IsLiveOut) {
+  auto *BBInfo = &BBInfos[BB];
+
+  if (IsLiveOut && BBInfo->LiveOutValue)
+    return BBInfo->LiveOutValue;
+
+  if (BBInfo->LiveInValue)
+    return BBInfo->LiveInValue;
+
+  SmallVector<BBValueInfo *, 4> DomPath = {BBInfo};
+  MachineBasicBlock *DomBB = BB, *TopDomBB = BB;
+  Register V;
+
+  while (DT.isReachableFromEntry(DomBB) && !DomBB->pred_empty() &&
+         (DomBB = DT.getNode(DomBB)->getIDom()->getBlock())) {
+    BBInfo = &BBInfos[DomBB];
+    if (BBInfo->LiveOutValue) {
+      V = BBInfo->LiveOutValue;
+      break;
+    }
+    if (BBInfo->LiveInValue) {
+      V = BBInfo->LiveInValue;
+      break;
+    }
+    TopDomBB = DomBB;
+    DomPath.emplace_back(BBInfo);
+  }
+
+  if (!V) {
+    V = createInst(TargetOpcode::IMPLICIT_DEF, TopDomBB,
+                   TopDomBB->getFirstNonPHI())
+            .getReg(0);
+  }
+
+  for (auto *BBInfo : DomPath)
+    // Loop above can insert new entries into the BBInfos map: assume the
+    // map shouldn't grow as the caller should have been allocated enough
+    // buckets, see [1].
+    BBInfo->LiveInValue = V;
+
+  return V;
+}
+
+/// Perform all the necessary updates, including new PHI-nodes insertion and the
+/// requested uses update.
+void MachineSSAUpdater2::calculate() {
+  MachineForwardIDFCalculator IDF(DT);
+
+  SmallPtrSet<MachineBasicBlock *, 2> DefBlocks;
+  for (auto [BB, V] : Defines)
+    DefBlocks.insert(BB);
+  IDF.setDefiningBlocks(DefBlocks);
+
+  SmallPtrSet<MachineBasicBlock *, 2> UsingBlocks(UseBlocks.begin(),
+                                                  UseBlocks.end());
+  SmallVector<MachineBasicBlock *, 32> IDFBlocks;
+  SmallPtrSet<MachineBasicBlock *, 32> LiveInBlocks;
```

Comment:
This is probably too big for a SmalLPtrSet

Reply:
Made 4 as default size.